### DISCUSSION TITLE
Pagination on nested fields (a post has a comment[])

### DISCUSSION THREAD

**Discussion by Knat-Dev** [urql-graphql/urql#1066] [Q&A] [answered]
First of all, I'm new to GraphQL and URQL, sorry if this is a noob question :)
I have a query that's supposed to get me one post from the db, and also paginate the comments for that post, there it is:

```
query Post($id: String!, $input: PaginationInput!) {
  post(id: $id) {
    id
    title
    text
    textSnippet
    commentCount
    voteStatus
    points
    createdAt
    updatedAt
    comments(input: $input) {
      comments {
        ...RegularComment
      }
      hasMore
    }
    creator {
      id
      username
    }
  }
}
```
The pagination input:
```
input PaginationInput {
  limit: Float!
  cursor: Float
}
```
RegularComment Fragment:
```
fragment RegularComment on Comment {
  id
  text
  createdAt
  creator {
    id
    username
  }
}
```

Right now when I try to fetch more comments the page reloads and not extra data is retrieved.
I did make pagination work for non-nested objects changing the SimplePagination function to meet my case (fetching paginated posts)

Is there a reason to keep going this way? or is there anything better?

Thank you in advance.

**Comment by frederikhors** [comment]
Can you create a reproduction on codesandbox, please?

**Reply by Knat-Dev** [reply]
I'll try to. (if you can tell me of an easy way I'll appreciate it)
I'm using Typescript+Type-GraphQL+Mongoose+Typegoose for my setup on my backend.
the problem I get is on my React frontend when updating the state that affects the query parameters.
The thing is I don't know how to update the cache for such a scenario.

**Reply by frederikhors** [reply]
You can use this for fake backend: https://graphqlzero.almansi.me/ and use your React code.

**Comment by Knat-Dev** [comment, marked as answer]
I managed to figure it out eventually.
I still had much to read on the docs, and it did help.

thanks anyways,
closable.

**Reply by JoviDeCroock** [reply]
Hey,

Let's post the solution, that way someone bumping into this discussion gets the correct information, I assume you used one of the pagination helpers or?

**Reply by frederikhors** [reply]
Yes, please, @Knat-Dev.

**Reply by Knat-Dev** [reply]
I did not read the information from the cache correctly, or from the right path. 
I had to put my pagination function inside the comments field of a Post object basically.

inside the creation of the client:
```
...
cacheExchange({
        resolvers: {
          // it is possible to reach inside the Post object and into it's comments field, fixed my problem
          Post: {
            comments: cursorPaginationComments(), // just a modified simplePagination() function, close to the one on the docs
          },
          Query: {
            posts: cursorPagination(),
          },
        },
...
```

there you go,
Thanks.

**Reply by Svarto** [reply]
@Knat-Dev I am facing the exact same challenge as you. Could you give some more details how you modified the simplePagination? Would be super helpful, thank you!

**Reply by kitten** [reply]
@Svarto Just to clarify, the problem here was that a pagination field was on `SomeType.field` rather than `Query.field` and the misunderstanding came from not knowing that resolvers (including the pagination helpers) can be placed on any type's fields rather than just the root `Query` type. So no further modification was required in this case

**Reply by Knat-Dev** [reply]
Sorry about the delay. 
there it is!

```typescript
const cursorPaginationComments = (): Resolver => {
  return (_parent, fieldArgs, cache, info) => {
    const { parentKey: entityKey, fieldName } = info;
    const allFields = cache.inspectFields(entityKey);
    const fieldInfos = allFields.filter((info) => info.fieldName === fieldName);
    const size = fieldInfos.length;
    if (size === 0) {
      return undefined;
    }

    const results: DataField[] = [];

    const fieldKey = `${fieldName}(${stringifyVariables(fieldArgs)})`;
    const isFieldKeyInCache = cache.resolve(
      cache.resolveFieldByKey(entityKey, fieldKey) as string,
      'comments'
    );
    info.partial = !isFieldKeyInCache;

    let hasMore = true;
    fieldInfos.forEach((fi) => {
      const key = cache.resolveFieldByKey(entityKey, fi.fieldKey) as string;
      const data = cache.resolve(key, 'comments');
      const _hasMore = cache.resolve(key, 'hasMore');
      if (!_hasMore) hasMore = _hasMore as boolean;
      results.push(...data);
    });

    return {
      __typename: 'PaginatedComments',
      hasMore,
      comments: results,
    };
  };
};
```

If you have any further questions, I'll try to help!

**Reply by nandorojo** [reply]
To clarify, should the pagination "just work" for nested fields as long as they receive the `limit` and `offset` arguments?

**Reply by RichardGrant-93** [reply]
uhg.. I hate that we have to write a custom resolver for nested pagination.

**Reply by nandorojo** [reply]
pretty sure you don’t. just do `field.nestedField: simplePagination()` right?

**Reply by JasperH8g** [reply]
> pretty sure you don’t. just do field.nestedField: simplePagination() right?

@nandorojo can you clarify on this a bit? How are we supposed to do this? Also, is `simplePagination` supposed to be compatible with nested fields in de result of the nested query as well? I can't find any info about this 🤔

**Reply by nandorojo** [reply]
So I was wrong here. I tried to do this to get a list of `eventById.rsvps`. However, it doesn't work.

```ts
eventById(parent, { id }) {
  return {
    __typename: 'Event',
    id,
    rsvps: simplePagination({
      limitArgument: 'limit',
      offsetArgument: 'offset',
    }),
  }
}
```

The query just never resolves when I add the `simplePagination` there.

**Reply by nandorojo** [reply]
Aha, I think this is how you do it:

```ts
resolvers: {
  Post: {
    comments: simplePagination()
  }
}
```

Notice that it isn't inside of `Query`. No need to write an exchange.
